src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2010 @@
+ */
+
+package scala.collection
+package immutable
+
+import scala.annotation.switch
+import scala.collection.mutable.{ArrayBuilder, Builder, ReusableBuilder}
+import scala.collection.generic.DefaultSerializable
+import scala.reflect.ClassTag
+import java.util.{Arrays, Spliterator}
+import java.util.Arrays.{copyOf, copyOfRange}
+import java.lang.Math.{abs, max => mmax, min => mmin}
+
+import NVectorStatics.{Arr1, Arr2, Arr3, Arr4, Arr5, Arr6}
+import scala.collection.Stepper.EfficientSplit
+
+
+object NVector extends StrictOptimizedSeqFactory[NVector] {
+  import NVectorStatics._
+
+  def empty[A]: NVector[A] = NVector0
+
+  def from[E](it: collection.IterableOnce[E]): NVector[E] =
+    it match {
+      case v: NVector[E] => v
+      case _ =>
+        val knownSize = it.knownSize
+        if (knownSize == 0) empty[E]
+        else if (knownSize > 0 && knownSize <= WIDTH) {
+          val a1 = new Array[Any](knownSize)
+          it.iterator.copyToArray(a1)
+          new NVector1[E](a1.asInstanceOf[Arr1])
+        } else {
+          (newBuilder ++= it).result()
+        }
+    }
+
+  def newBuilder[A]: ReusableBuilder[A, NVector[A]] = new NVectorBuilder[A]
+
+  /** Create a Vector with the same element at each index.
+    *
+    * Unlike `fill`, which takes a by-name argument for the value and can thereby
+    * compute different values for each index, this method guarantees that all
+    * elements are identical. This allows sparse allocation in O(log n) time and space.
+    */
+  def fillSparse[A](n: Int)(elem: A): NVector[A] = {
+    if(n <= 0) NVector0
+    else {
+      val b = new NVectorBuilder[A]
+      b.initSparse(n, elem)
+      b.result()
+    }
+  }
+}
+
+
+/** Vector is a general-purpose, immutable data structure.  It provides random access and updates
+  * in O(log n) time, as well as very fast append/prepend/tail/init (amortized O(1), worst case O(log n)).
+  * Because vectors strike a good balance between fast random selections and fast random functional updates,
+  * they are currently the default implementation of immutable indexed sequences.
+  *
+  * Vectors are implemented by radix-balanced finger trees of width 32. There is a separate subclass
+  * for each level (0 to 6, with 0 being the empty vector and 6 a tree with a maximum width of 64 at the
+  * top level).
+  *
+  * Tree balancing:
+  * - Only the first dimension of an array may have a size < WIDTH
+  * - In a `data` (central) array the first dimension may be up to WIDTH-2, in a prefix or suffix WIDTH-1
+  * - prefix1 and suffix1 are never empty
+  * - Balancing does not cross the main data array (i.e. prepending never touches the suffix and appending never touches
+  *   the prefix). The level is increased/decreased when the affected side plus main data is already full/empty
+  * - All arrays are left-aligned and truncated
+  *
+  * In addition to the data slices (`prefix`, `prefix2`, ..., `dataN`, ..., `prefix2`, `prefix1`) we store a running
```

Comment:
Is there any reference we may cite, regarding the design and implementation of radix-balanced finger trees? For instance, is [this ICFP](https://dblp.uni-trier.de/rec/bibtex/journals/pacmpl/Puente17) article closely related?

Reply:
As I mentioned above I am not aware of any pre-existing references. The Bolivar paper (I found a PDF [here](https://public.sinusoid.es/misc/immer/immer-icfp17.pdf)) covers RRB trees, like the two papers I referenced in the PR description. My implementation uses non-relaxed trees. Relaxed trees can greatly improve insertion and concatenation performance but will show worse results for indexed access. I think it would be interesting to see if RBF trees can be generalized to RRBF trees. I expect it to be possible in principle but with higher implementation complexity than for a relaxed version of normal RB trees.

The "Off-Tree Tail" and "Left Shadow" optimizations from section 2.3 are worth mentioning in this context. The fingers are a generalization of an off-tree tail. While the off-tree tail is only the last one-dimensional slice, a finger extends this to all dimensions from 1 to n-1 for an n-dimensional vector. The left shadow is implemented by the old Vector. The finger on the left now provides the same advantages with a more compact memory representation.

Reply:
Should this be:

```
(`prefix`, `prefix2`, ..., `dataN`, ..., `suffix2`, `suffix1`)
```
?

Reply:
No, it should be `(`prefix1`, `prefix2`, ..., `dataN`, ..., `suffix2`, `suffix1`)`. It got 3 of 5 names wrong...